xds/src/main/java/io/grpc/xds/XdsClientImpl.java

```diff
@@ -1444,126 +1639,263 @@ private void cleanUp() {
       }
-      DiscoveryRequest request =
-          DiscoveryRequest
+      DiscoveryRequestData request =
+          DiscoveryRequestData
               .newBuilder()
               .setVersionInfo(versionInfo)
-              .setNode(node.toEnvoyProtoNodeV2())
+              .setNode(node)
               .addAllResourceNames(resourceNames)
-              .setTypeUrl(typeUrl)
+              .setTypeUrl(resourceType)
               .setResponseNonce(nonce)
               .setErrorDetail(
                   com.google.rpc.Status.newBuilder()
                       .setCode(Code.INVALID_ARGUMENT_VALUE)
-                      .setMessage(message))
+                      .setMessage(message)
+                      .build())
               .build();
-      requestWriter.onNext(request);
-      logger.log(XdsLogLevel.DEBUG, "Sent NACK request\n{0}", request);
+      sendDiscoveryRequest(request);
+    }
+  }
+
+  private final class AdsStreamV2 extends AbstractAdsStream {
+    private final io.envoyproxy.envoy.service.discovery.v2.AggregatedDiscoveryServiceGrpc
+        .AggregatedDiscoveryServiceStub stubV2;
+    private StreamObserver<io.envoyproxy.envoy.api.v2.DiscoveryRequest> requestWriterV2;
+
+    AdsStreamV2(io.envoyproxy.envoy.service.discovery.v2.AggregatedDiscoveryServiceGrpc
+        .AggregatedDiscoveryServiceStub stubV2) {
+      this.stubV2 = checkNotNull(stubV2, "stubV2");
+    }
+
+    @Override
+    void start() {
+      StreamObserver<io.envoyproxy.envoy.api.v2.DiscoveryResponse> responseReaderV2 =
+          new StreamObserver<io.envoyproxy.envoy.api.v2.DiscoveryResponse>() {
+            @Override
+            public void onNext(io.envoyproxy.envoy.api.v2.DiscoveryResponse response) {
+              DiscoveryResponseData responseData =
+                  DiscoveryResponseData.fromEnvoyProtoV2(response);
+              if (logger.isLoggable(XdsLogLevel.DEBUG)) {
+                logger.log(
+                    XdsLogLevel.DEBUG,
+                    "Received {0} response:\n{1}",
+                    responseData.getResourceType(),
+                    respPrinter.print(response));
+              }
+              onDiscoveryResponse(responseData);
+            }
+
+            @Override
+            public void onError(Throwable t) {
+              AdsStreamV2.this.onError(t);
+            }
+
+            @Override
+            public void onCompleted() {
+              AdsStreamV2.this.onCompleted();
+            }
+          };
+      requestWriterV2 = stubV2.withWaitForReady().streamAggregatedResources(responseReaderV2);
+    }
+
+    @Override
+    void sendDiscoveryRequest(DiscoveryRequestData request) {
+      checkState(requestWriterV2 != null, "ADS stream has not been started");
+      io.envoyproxy.envoy.api.v2.DiscoveryRequest requestProto =
+          request.toEnvoyProtoV2();
+      requestWriterV2.onNext(requestProto);
+      logger.log(XdsLogLevel.DEBUG, "Sent ACK request\n{0}", requestProto);
```

Comment:
Wrong log. The actual one should have already been logged. Same for the v3 class.

Reply:
Good catch. Hmm.. it's quite annoying for the logging. We need log the versioned proto message actually being sent out, so we are not able to log it inside send{Xds,Ack,Nack}Request(DiscoveryRequestData).

Reply:
I mean you can just log in the method sendXdsRequest/sendAckRequest/sendNackRequest after the line of calling `sendDiscoveryRequest`.

Reply:
> just log in the method sendXdsRequest/sendAckRequest/sendNackRequest after the line of calling sendDiscoveryRequest.

Here we want to log the actual proto message sent out. Inside `send{Xds,Ack,Nack}Request(DiscoveryRequestData)`, you can only log `DiscoveryRequestData.toString()` (which is not necessary, because most of the data was already logged in `sendXdsRequest` "Sending XXX request for resources: {0}"), unless `sendDiscoveryRequest` returns the proto message.

Reply:
You can have one log message in sendAckRequest/sendNackRequest indicating if the request is ACK or NACK. Then another log message just with the proto without description in sendDiscoveryRequest. Do not add the AckType enum, it complicates things for little gain.

Reply:
Done.

Reply:
Can you preserve the messages:

```
"Sending ACK/NACK"
proto
```

Most people are eyeballing for "ACK" or "NACK" keywords.

Reply:
Sounds good. Done.